ntbs-service/Services/AzureAdDirectoryService.cs

```diff
@@ -38,20 +40,20 @@ public AzureAdDirectoryService(IGraphServiceClient graphServiceClient, AdOptions
             _adOptions = adOptions;
         }
 
-        public void Dispose()
+        public async Task<IList<(Models.Entities.User user, IList<TBService> tbServicesMatchingGroups)>>
+            LookupUsers(IList<TBService> tbServices)
         {
-
-        }
-
-        public async Task<IEnumerable<(Models.Entities.User user, List<TBService> tbServicesMatchingGroups)>> LookupUsers(IList<TBService> tbServices)
-        {
-            IList<(Models.Entities.User user, List<TBService> tbServicesMatchingGroups)> userWithTbServiceGroups = new List<(Models.Entities.User user, List<TBService> tbServicesMatchingGroups)>();
+            var userWithTbServiceGroups =
+                new List<(Models.Entities.User user, IList<TBService> tbServicesMatchingGroups)>();
             var graphUsers = await GetAllDirectoryEntries();
 
-            // ensure that users are unique
-            graphUsers = graphUsers.GroupBy(u => u.UserPrincipalName).Select(u => u.FirstOrDefault());
+            // Ensure that users are unique
+            var uniqueUsers = graphUsers
+                .GroupBy(user => user.UserPrincipalName)
+                .Select(grouping => grouping.FirstOrDefault())
+                .Where(user => user != null);
 
-            foreach (var user in graphUsers)
+            foreach (var user in uniqueUsers)
```

Comment:
You might be able to get away with a `Task.WhenAll` rather than a `foreach` here, as the Graph API requests should be able to run in parallel. However, if this is tried and tested then may be better not to risk it.

Reply:
Yeah that's worth doing. I've just tested this locally so and it works great. Making good use of `Task.WhenAll` in other places has made the user sync job noticeably faster, I must have missed this one so thanks!